### ISSUE TITLE
`Astro.Solar.utc_sun_position/4` for any solar elevation different than 90 returns incorrect results

### ISSUE THREAD

**Issue by cloud8421** [kipcole9/astro#5]
Hello,

Thanks a lot for your work, and apologies in advance, I'm not at all an expert into the field.

I'm trying to calculate dawn and dusk for a given location, and thought I could use `Astro.Solar.utc_sun_position/4`.

After some initial tries, I couldn't get correct results (compared against some online sources) so I started spelunking the implementation. After some debugging, I ended up starting to play with:

https://github.com/kipcole9/astro/blob/9a450178310679452e6e6b5ccf88be0d9cbfade0/lib/astro/earth.ex#L113-L119

Is it possible that the fallback implementation is incorrect and should return the `solar_elevation` that is passed to the function?

Otherwise any call to `utc_sun_position/4` that doesn't use `90.0` degrees returns a hardcoded value.

For comparison, I checked other implementations in other libraries, and it seems that use a similar approach, and pass on the provided value, e.g.

https://github.com/pinnymz/ruby-zmanim/blob/8e7b4a4ea96e8859258026d2bd310025ef71b513/lib/zmanim/util/astronomical_calculations.rb#L29

Again, really not my domain. But if I make the suggested change, I get consistent values. Here's some test code:

```
current_date = Date.utc_today()
current_time = DateTime.utc_now() |> DateTime.shift_zone!("Europe/London")
crouch_end_z = %Geo.PointZ{coordinates: {-0.1062, 51.5171, 41.0}}

{:ok, dawn_hours_since_midnight} =
  Astro.Solar.utc_sun_position(current_time, crouch_end_z, 96.0, :sunrise)

{:ok, dusk_hours_since_midnight} =
  Astro.Solar.utc_sun_position(current_time, crouch_end_z, 96.0, :sunset)

{:ok, utc_dawn_datetime} = Astro.Time.moment_to_datetime(dawn_hours_since_midnight, current_date)
{:ok, utc_dusk_datetime} = Astro.Time.moment_to_datetime(dusk_hours_since_midnight, current_date)

utc_dawn_datetime
|> DateTime.shift_zone!("Europe/London")
|> IO.inspect(label: "Dawn at Crouch End")

utc_dusk_datetime
|> DateTime.shift_zone!("Europe/London")
|> IO.inspect(label: "Dusk at Crouch End")
```

**Comment by kipcole9** [comment]
Wow, thanks very much for taking the time to look into this.  I'm no expert either - been a total learning experience for me - so I greatly appreciate the assist.

I'll dive into this in a few hours when the sun comes up (ha ha).

**Comment by kipcole9** [comment]
I've pushed [a commit](https://github.com/kipcole9/astro/commit/4866dcab0c911903ff0c5834ef9fb277681de2a8) that applies the bug fix you suggested. I added two test cases so far to validate and they are close to online sources but not the same. Around 1 minute 50 seconds difference on these two test cases. I note that on the test data for sunrise/sunset with solar elevation 90 degrees (which then corrects for refraction etc) the calculated results are correct to the minute compared with online sources. I wonder if online sources are apply refraction even on civil, nautical and astronomical solar elevations?

Anyway, here are the two tests so far. Are these the results you expect? Note that I'm using the public API in `Astro`:

```elixir
    test "Crouch End dawn" do
      crouch_end_z = %Geo.PointZ{coordinates: {-0.1062, 51.5171, 41.0}}
      {:ok, expected_date_time} = DateTime.new(~D[2024-05-26], ~T[04:09:50.000000], "Europe/London")

      assert {:ok, ^expected_date_time} =
        Astro.sunrise(crouch_end_z, ~D[2024-05-26], solar_elevation: :civil)
    end

    test "London dawn" do
      london_z = %Geo.PointZ{coordinates: {-0.1276, 51.5072, 11.0}}
      {:ok, expected_date_time} = DateTime.new(~D[2024-05-26], ~T[04:09:59.000000], "Europe/London")

      assert {:ok, ^expected_date_time} =
        Astro.sunrise(london_z, ~D[2024-05-26], solar_elevation: :civil)
    end
```

**Comment by kipcole9** [comment]
I made an [additional commit](https://github.com/kipcole9/astro/commit/f4a0e5d81dde75c2119dfdd1f60f29249bb4619f) that applies elevation adjustment even in cases where the solar elevation isn't 90 degrees. 

Results now appear consistent with online sources to the minute (truncated, not rounded).

Hopefully this is getting close to expectations?

```elixir
    test "Crouch End dawn" do
      crouch_end_z = %Geo.PointZ{coordinates: {-0.1062, 51.5171, 41.0}}

      {:ok, expected_date_time} =
        DateTime.new(~D[2024-05-26], ~T[04:07:56.000000], "Europe/London")

      assert {:ok, ^expected_date_time} =
        Astro.sunrise(crouch_end_z, ~D[2024-05-26], solar_elevation: :civil)
    end

    test "London dawn" do
      london_z = %Geo.PointZ{coordinates: {-0.1276, 51.5072, 21.0}}

      {:ok, expected_date_time} =
         DateTime.new(~D[2024-05-26], ~T[04:08:37.000000], "Europe/London")

      assert {:ok, ^expected_date_time} =
        Astro.sunrise(london_z, ~D[2024-05-26], solar_elevation: :civil)
    end
```

**Comment by cloud8421** [comment]
Thank you @kipcole9 for the quick turnaround!

I'll test the branch - I would also assume that refraction correction is needed now that you mention it.

Precision to the minute is absolutely good enough for me - I'm not sure about other use cases.

I'm writing a custom firmware for https://shop.pimoroni.com/products/mood-light-pi-zero-w-project-kit?variant=38477389450 to change the color of the light depending on the time of day, so minute precision is absolutely fine. Thanks again! I'll report back in this issue.

**Comment by kipcole9** [comment]
The literature suggests that refraction isn’t calculated for anything except 90 degrees (ie sunrise). I did try adding it in but the results were then not the same as online sources. 

Look forward to your observations - and I really hope it’s correct now. Debugging this stuff is really hard :-(. But still fun,

**Comment by cloud8421** [comment]
> The literature suggests that refraction isn’t calculated for anything except 90 degrees (ie sunrise). I did try adding it in but the results were then not the same as online sources.

Ah yeah, sorry - you're right. I see you adjusted elevation, which makes more sense.

My tests are great with the additional commits - I'm happy to keep pulling from master until a new release is cut, but no rush from my perspective. Thanks!

**Comment by kipcole9** [comment]
I've published [Astro version 1.0.2](https://hex.pm/packages/astro) with the following changelog entry:

### Bug Fixes

* Fix sunrise/sunset calculations when the solar elevation isn't 90 degrees. Thanks to @cloud8421 for the issue. Closes #5.
